Utilities/XrdAdaptor/src/XrdRequestManager.cc

```diff
@@ -839,17 +839,38 @@ void RequestManager::requestFailure(std::shared_ptr<XrdAdaptor::ClientRequest> c
   m_disabledSources.insert(source_ptr);
 
   std::unique_lock<std::recursive_mutex> sentry(m_source_mutex);
-  if ((!m_activeSources.empty()) && (m_activeSources[0].get() == source_ptr.get())) {
+  // Remove the failed source from the container of active sources
+  if (auto found = std::ranges::find_if(
+          m_activeSources, [&source_ptr](const std::shared_ptr<Source> &src) { return src.get() == source_ptr.get(); });
+      found != m_activeSources.end()) {
     auto oldSources = m_activeSources;
-    m_activeSources.erase(m_activeSources.begin());
-    reportSiteChange(oldSources, m_activeSources);
-  } else if ((m_activeSources.size() > 1) && (m_activeSources[1].get() == source_ptr.get())) {
-    auto oldSources = m_activeSources;
-    m_activeSources.erase(m_activeSources.begin() + 1);
+    m_activeSources.erase(found);
     reportSiteChange(oldSources, m_activeSources);
   }
+  // Find a new source to send the request to
+  // - First, if there is another active source, use it
+  // - Then, if there are no active sources, if there are inactive
+  //   sources, use the best inactive source
+  // - Then, if there are no active or inactice sources, try to open a
+  //   new connection for a new source
   std::shared_ptr<Source> new_source;
-  if (m_activeSources.empty()) {
+  if (not m_activeSources.empty()) {
+    new_source = m_activeSources[0];
+  } else if (not m_inactiveSources.empty()) {
+    // similar logic as in checkSourcesImpl()
+    // assume the "sort open delay" doesn't matter in case of a request failure
+    auto bestInactiveSource =
+        std::min_element(m_inactiveSources.begin(),
```

Comment:
Is the best quality really the lowest number?

Reply:
As far as I can tell, yes. See e.g. existing code
https://github.com/cms-sw/cmssw/blob/eb45085aa953e72e9b379bb0b543a2897998809e/Utilities/XrdAdaptor/src/XrdRequestManager.cc#L455-L460

There is some discussion about the quality metric in https://github.com/cms-sw/cmssw/blob/master/Utilities/XrdAdaptor/doc/multisource_algorithm_design.txt
and _maybe_ it is computed in https://github.com/cms-sw/cmssw/blob/master/Utilities/XrdAdaptor/src/QualityMetric.cc, but I haven't tried to understand what is done there.

Reply:
OK, thanks, sounds good! :+1: